rust/src/conflict_checker.rs

```diff
@@ -0,0 +1,431 @@
+        self.check_no_metadata_updates()?;
+        self.check_for_added_files_that_should_have_been_read_by_current_txn()?;
+        self.check_for_deleted_files_against_current_txn_read_files()?;
+        self.check_for_deleted_files_against_current_txn_deleted_files()?;
+        self.check_for_updated_application_transaction_ids_that_current_txn_depends_on()?;
+        todo!()
+    }
+
+    /// Asserts that the client is up to date with the protocol and is allowed
+    /// to read and write against the protocol set by the committed transaction.
+    fn check_protocol_compatibility(&self) -> Result<(), CommitConflictError> {
+        for p in self.winning_commit_summary.protocol() {
+            if self.state.min_reader_version() < p.min_reader_version
+                || self.state.min_writer_version() < p.min_writer_version
+            {
+                return Err(CommitConflictError::ProtocolChanged);
+            };
+        }
+        if !self.winning_commit_summary.protocol().is_empty()
+            && self
+                .current_transaction_info()
+                .actions
+                .iter()
+                .any(|a| matches!(a, Action::protocol(_)))
+        {
+            return Err(CommitConflictError::ProtocolChanged);
+        };
+        Ok(())
+    }
+
+    /// Check if the committed transaction has changed metadata.
+    fn check_no_metadata_updates(&self) -> Result<(), CommitConflictError> {
+        // Fail if the metadata is different than what the txn read.
+        if !self.winning_commit_summary.metadata_updates().is_empty() {
+            Err(CommitConflictError::MetadataChanged)
+        } else {
+            Ok(())
+        }
+    }
+
+    /// Check if the new files added by the already committed transactions
+    /// should have been read by the current transaction.
+    fn check_for_added_files_that_should_have_been_read_by_current_txn(
+        &self,
+    ) -> Result<(), CommitConflictError> {
+        // Fail if new files have been added that the txn should have read.
+        let added_files_to_check = match self.isolation_level {
+            IsolationLevel::WriteSerializable
+                if !self.current_transaction_info().metadata_changed() =>
+            {
+                // don't conflict with blind appends
+                self.winning_commit_summary.changed_data_added_files()
+            }
+            IsolationLevel::Serializable | IsolationLevel::WriteSerializable => {
+                let mut files = self.winning_commit_summary.changed_data_added_files();
+                files.extend(self.winning_commit_summary.blind_append_added_files());
+                files
+            }
+            IsolationLevel::SnapshotIsolation => vec![],
+        };
+        // TODO here we need to check if the current transaction would have read the
+        // added files. for this we need to be able to evaluate predicates. Err on the safe side is
+        // to assume all files match
```

Comment:
What would you think of keeping this behavior if datafusion feature isn't included? That way we could still have the option of not including datafusion. I think there are some use cases where we don't want to require it as a dependency.

Reply:
Absolutely - I was contemplating how to do this for a bit and parsing complex predicates would certainly benefit from some more advanced parsing. However right now we only have min max and nulls anyhow, which the current predicate parsing should be able to parse as well.

When we apply the filters my feeling is that we want to go the arrow route - i.e. put all files with bounds in  a record batch and use arrow kernels. I'll see how far I get using "just" arrow and as you said, falling back to the most conservative case is always an option.